pkg/query-service/app/clickhouseReader/reader.go

```diff
@@ -1443,46 +1467,169 @@ func String(length int) string {
+		tagMapType := ""
+		switch item.(type) {
+		case model.TagQueryString:
+			tagMapType = constants.StringTagMapCol
+		case model.TagQueryNumber:
+			tagMapType = constants.NumberTagMapCol
+		case model.TagQueryBool:
+			tagMapType = constants.BoolTagMapCol
+		default:
+			// type not supported error
+			return nil, &model.ApiError{Typ: model.ErrorBadData, Err: fmt.Errorf("type not supported")}
+		}
+		switch item.GetOperator() {
+		case model.EqualOperator:
+			args = addArithmeticOperator(item, query, tagMapType, args, "=")
+		case model.NotEqualOperator:
+			args = addArithmeticOperator(item, query, tagMapType, args, "!=")
+		case model.LessThanOperator:
+			args = addArithmeticOperator(item, query, tagMapType, args, "<")
+		case model.GreaterThanOperator:
+			args = addArithmeticOperator(item, query, tagMapType, args, ">")
+		case model.InOperator:
+			args = addInOperator(item, query, tagMapType, args, false)
+		case model.NotInOperator:
+			args = addInOperator(item, query, tagMapType, args, true)
+		case model.LessThanEqualOperator:
+			args = addArithmeticOperator(item, query, tagMapType, args, "<=")
+		case model.GreaterThanEqualOperator:
+			args = addArithmeticOperator(item, query, tagMapType, args, ">=")
+		case model.ContainsOperator:
+			args = addContainsOperator(item, query, tagMapType, args, false)
+		case model.NotContainsOperator:
+			args = addContainsOperator(item, query, tagMapType, args, true)
+		case model.StartsWithOperator:
+			args = addStartsWithOperator(item, query, tagMapType, args, false)
+		case model.NotStartsWithOperator:
+			args = addStartsWithOperator(item, query, tagMapType, args, true)
+		case model.ExistsOperator:
+			args = addExistsOperator(item, query, tagMapType, args, false)
+		case model.NotExistsOperator:
+			args = addExistsOperator(item, query, tagMapType, args, true)
+		default:
+			return nil, &model.ApiError{Typ: model.ErrorExec, Err: fmt.Errorf("Tag Operator %s not supported", item.GetOperator())}
 		}
 	}
 	return args, nil
 }
 
-func (r *ClickHouseReader) GetTagFilters(ctx context.Context, queryParams *model.TagFilterParams) (*[]model.TagFilters, *model.ApiError) {
+func addInOperator(item model.TagQuery, query *string, tagMapType string, args []interface{}, not bool) []interface{} {
+	values := item.GetValues()
+	notStr := ""
+	if not {
+		notStr = "NOT"
+	}
+	for i, value := range values {
+		tagKey := "inTagKey" + String(5)
+		tagValue := "inTagValue" + String(5)
+		if i == 0 && i == len(item.GetValues())-1 {
+			*query += fmt.Sprintf(" AND %s %s[@%s] = @%s", notStr, tagMapType, tagKey, tagValue)
+		} else if i == 0 && i != len(item.GetValues())-1 {
+			*query += fmt.Sprintf(" AND %s (%s[@%s] = @%s", notStr, tagMapType, tagKey, tagValue)
+		} else if i != 0 && i == len(item.GetValues())-1 {
+			*query += fmt.Sprintf(" OR %s[@%s] = @%s)", tagMapType, tagKey, tagValue)
+		} else {
+			*query += fmt.Sprintf(" OR %s[@%s] = @%s", tagMapType, tagKey, tagValue)
+		}
```

Comment:
The mixed use of the pointer and regular types makes it harder to follow. If I understand this logic correctly, this takes into account what operator to use because it is combined with other key-value sub filter queries. There is some esoteric logic here. You could change each of these filter items to return the sub query and, in the caller, join them by the appropriate operator.

Reply:
ok 👍